src/controllers/orderController.js

```diff
@@ -162,4 +164,45 @@ const deleteOrder = async (req, res) => {
   */
 };
 
-module.exports = { createOrder, getAllOrders, getSingleOrder, deleteOrder };
+const handleStripePayment = async (req, res) => {
+  // Receive Stripe Signature
+  const stripeSignature = req.headers['stripe-signature'];
+
+  //  Construct the Webhook Event
+  const event = stripe.webhooks.constructEvent(
+    req.body, // Raw text body payload received from Stripe
+    stripeSignature, // Value of the `stripe-signature` header from Stripe
+    process.env.STRIPE_CLI_WEBHOOK_SECRET // Webhook Signing Secret
+  );
+
+  //  Retrieve Payment Intent and Order Information
+  const paymentIntent = event.data.object; // Payment intent information
+  const orderId = paymentIntent.metadata.orderId; // Order ID passed in metadata
+  // Additional order-related info if needed
+
+  //  Update Order Status
+  const order = await Order.findById(orderId);
+
+  if (event.type === 'payment_intent.succeeded') {
+    // Update order status to successful
+    order.orderStatus = 'payment_successful';
+  } else if (event.type === 'payment_intent.payment_failed') {
+    // Update order status to failed
+    order.orderStatus = 'payment_failed';
+  }
+
+  // Save the updated order
+  await order.save();
+
+  // Handle other actions based on the event type if needed
+
+  return res.status(200);
```

Comment:
Another part I missed here earlier; `res.status(...)` is not enough to tell Express to send the response. It waits to know what _body_ to send if any.

Since Stripe doesn't care about the response body, just the status, we can send an empty response like this: `res.status(200).end()`

(`res.status(200).json({})` or `res.status(200).send()` also would work though those are technically sending a non-empty response body, with either just a blank json object in the first case, or an empty string response body in the second case).

Reply:
@akosasante , Akos,  why- if our `event.type === **'payment_intent.succeeded'` - and this is what we do receive after making a payment + the `orderStatus `in mongo is being changed to `payment successful`- but  why in stripe dashboard we  still see 
that webhook delivery failed- once it returned timed out error- which i was trying to resolve with adding await to 
```
 const event =  stripe.webhooks.constructEvent(
    req.body, // Raw text body payload received from Stripe
    stripeSignature, // Value of the `stripe-signature` header from Stripe
    process.env.STRIPE_CLI_WEBHOOK_SECRET // Webhook Signing Secret
  );
```
in handleSptipePayment controller  but it didn't helped- and now I get 404 - besides I've added empty body to  `res.status(200).end()`
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/60b5ab29-cbbd-4bed-86bd-fc3e4052ccb0)
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/41b6e50d-14a2-47e5-889e-f22c11a7a541)

Reply:
Hi @NatachaKey I just tried with your latest changes and things seem to be working well.

One thing I didn't realize, as I'm a bit new to Stripe development as well, is that the Webhooks table you've screenshotted, will only tell us things from the perspective of the live website. And since we haven't merged this PR yet, of course, the live website does not yet have a `/api/v1/orders/payment_status` endpoint, which is what is causing the "Route not found" 404 errors seen there. I thought that forwarding to the CLI meant that it would _only_ call the localhost version, but it looks like it actually copies the request so it's calling both the production music app server AND your local one. 

The main thing we are concerned about here at the moment before merging is that the code works as expected when we forward the call to our local servers.

Here are the steps I took to test:

- Frontend running `PS-81-create-sidebar-for-order-details` branch
- Backend running this `stripeWebhook` branch
- Another terminal with Stripe CLI running and forwarding any calls made to the live webhook endpoint: `D:\stripe\stripe.exe listen --load-from-webhooks-api --forward-to localhost:8000`
- Created an order by opening `localhost:3000`, adding an album to cart and clicking "Proceed to Checkout"
  - At this point, verified from backend logs that a POST request was made to `/api/v1/orders` ✔️ 
r
  - Verified that the Mongo database had an order created with `orderStatus` being `pending` and a `paymentId` populated ✔️ 
  - Verified in Stripe developer dashboard that an incomplete payment was started for the order: https://dashboard.stripe.com/test/payments ✔️ 
  - In my terminal that's running the Stripe CLI, I could see that the Stripe servers had received a `payment_intent.created`  call ✔️ 
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/9094098/324a0bac-6be2-4982-9270-74eba24a34ad)

- Submitted a valid payment for the order on the checkout page using a Stripe test credit card number
  - Verified the backend logs that a POST request was made to` /api/v1/orders/payment_status` and that 200 response was returned ✔️ 
  - Verified in Mongo database that the same order had its `orderStatus` updated to `payment_successful` ✔️
  - Verified in Stripe CLI that it tool had made a request to the correct endpoint and received a 200 ✔️
  - Verified in Stripe dashboard in the **EVENTS** section of the Developers page, that the response had been successfully received in the CLI: https://dashboard.stripe.com/test/events/evt_3O1u3CKnb0YvaiB30lebEI7X ✔️ 


If the above steps work for you as well, then you can feel free to merge!

Reply:
@akosasante got it!! thank you so much!!! You are the BEST!!!😍
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/a271311c-facb-4f73-a433-d0f5495eebf4)
just applied formatting and linting -not sure why the CI jobs fail here - I've checked formatting several times locally and when I push it says that there's nothing to commit..
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/c449148c-655a-4ae2-b913-a85e4668de49)

Reply:
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/da13cee9-9226-473a-a0fe-40c4cd7d2c65)

Reply:
:tada: 

---

That's odd, when I pull this branch and run the prettier check I get:

```
aasante@aasante-C02F75HHMD6R backend % git switch stripeWebhook
Switched to branch 'stripeWebhook'
Your branch is behind 'origin/stripeWebhook' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
aasante@aasante-C02F75HHMD6R backend % git pull
Updating 62806f8..b75d636
Fast-forward
 src/controllers/orderController.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

aasante@aasante-C02F75HHMD6R backend % npm run format:check

> bb-practicum-team-4-back@1.0.0 format:check
> prettier --check .

Checking formatting...
[warn] src/controllers/orderController.js
[warn] Code style issues found in the above file. Run Prettier to fix.
```

I wonder if updating your branch (either in Github using the button, and then do `git pull` OR in your terminal by doing `git fetch origin main && git merge origin/main`) . And then running the format + lint fix commands maybe will work?

---

Ohhh interesting that last screenshot might help us fix husky issue!

typicode/husky#944

Try the following:

*   copy the content of `.husky/pre-commit` into another file for safekeeping
*   run `npx husky-init` (unfortunately this will overwrite the `pre-commit` file)
    *   Check that there is a `.husky/_/husky.sh` file existing now
*   paste in the contents that were in the `pre-commit` file before back into it
*   try testing husky again the way we were talking about in the other PR (make a bad syntax change, try to add and commit it and see if `lint-staged` runs).

Reply:
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/39ff3d89-2a6f-40d0-b777-32b7978b6d37)
oh thank you, finally  able to merge it

Reply:
Awesome!! So glad we got it working 🎉